# 1. Import Libraries

In [ ]:
import numpy as np
import pandas as pd
import datetime
import time
from plotly import __version__
from plotly.offline import download_plotlyjs, init_notebook_mode, plot, iplot
import plotly.express as px
print(__version__) # requires version >= 1.9.0
import cufflinks as cf
init_notebook_mode(connected=True)
cf.go_offline()


from pymongo import MongoClient
%matplotlib inline

5.5.0


In [ ]:
pd.set_option('display.max_columns', None)

---
# 2. Connect to MongoDB Compass

In [ ]:
# Instantiate MongoDB client by localhost 27017
client = MongoClient("mongodb://localhost:27017")

# Get database logee_trans
db = client.get_database("logee_trans")

# Retrieve collection deliveryDetails from database
deliv_details = db.get_collection("deliveryDetails")

---
# 3. Convert Data
Convert the retrieved .json format data into pandas dataframe.

In [ ]:
df_delivery = pd.DataFrame(list(deliv_details.find({})))

In [ ]:
print(df_delivery.info())
print(df_delivery.shape)
display(df_delivery.head(20))

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13737 entries, 0 to 13736
Data columns (total 99 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   _id                            13737 non-null  object 
 1   orderNumber                    13728 non-null  object 
 2   originLatitude                 13728 non-null  object 
 3   destinationAddress             13728 non-null  object 
 4   distance                       13728 non-null  object 
 5   originAddress                  13728 non-null  object 
 6   itemHeight                     5623 non-null   float64
 7   itemValue                      13728 non-null  object 
 8   destinationLocation            13728 non-null  object 
 9   vehicleGroupName               13728 non-null  object 
 10  itemCategoryId                 13728 non-null  object 
 11  itemName                       13728 non-null  object 
 12  pickupTime                     13728 non-null 

_id      orderNumber      originLatitude  \
0   5f151562f2a56c0010dd128f  LT-202007200002          -6.2382699   
1   5f437721351da9001004d1e0  LT-202008240003          -6.1021109   
2   5f509dbf351da9001004d220  LT-202009030014          -6.2167736   
3   5f50a5b8351da9001004d226  LT-202009030017          -6.3227303   
4   5f563459351da9001004d259  LT-202009070029          -6.1656824   
5   5f575e73351da9001004d261  LT-202009080033          -6.1656824   
6   5f58edeb351da9001004d271  LT-202009090041          -6.1403479   
7   5f5df2e9c444250011e5e7c4  LT-202009130057          -6.2195686   
8   5f6054bec444250011e5e810  LT-202009150078          -6.5322491   
9   5f2a74cff2a56c0010dd12b3  LT-202008050002          -6.3227303   
10  5eff1a56e58bc77f11efba9b              NaN                 NaN   
11  5eff1a56e58bc77f11efba9c              NaN                 NaN   
12  5f15134df2a56c0010dd128d  LT-202007200001  -7.005740899999999   
13  5f4f648b351da9001004d21c  LT-202009020012          -6.1403479   
14  5f521865351da9001004d234  LT-202009040024          -6.1486651   
15  5f5dc0ebc444250011e5e7b1  LT-202009130054           3.5258916   
16  5f604c6fc444250011e5e804  LT-202009150074          -7.2948875   
17  5f63439bc444250011e5e840  LT-202009170093          -6.1826294   
18  5f634a55c444250011e5e849  LT-202009170096          -6.3015658   
19  5f6495a1c444250011e5e85b  LT-202009180101          -6.2167736   

                                   destinationAddress distance  \
0   Tanjung Priok, North Jakarta City, Jakarta, In...     28.0   
1   Kapuk Muara, Kec. Penjaringan, Kota Jkt Utara,...     28.1   
2   Tj. Priok, Kota Jkt Utara, Daerah Khusus Ibuko...     52.6   
3   Tj. Priok, Kota Jkt Utara, Daerah Khusus Ibuko...     75.6   
4   Tj. Priok, Kota Jkt Utara, Daerah Khusus Ibuko...     29.4   
5   Tj. Priok, Kota Jkt Utara, Daerah Khusus Ibuko...     29.4   
6   Tj. Priok, Kota Jkt Utara, Daerah Khusus Ibuko...     18.2   
7   Kec. Danurejan, Kota Yogyakarta, Daerah Istime...      559   
8   Tj. Priok, Kota Jkt Utara, Daerah Khusus Ibuko...     51.8   
9   North Jakarta, North Jakarta City, Jakarta, In...     71.3   
10                                                NaN      NaN   
11                                                NaN      NaN   
12  Tanjung Priok, North Jakarta City, Jakarta, In...      361   
13  Tj. Priok, Kota Jkt Utara, Daerah Khusus Ibuko...     18.2   
14  Tj. Priok, Kota Jkt Utara, Daerah Khusus Ibuko...     25.4   
15  Kec. Senen, Kota Jakarta Pusat, Daerah Khusus ...     1899   
16        Kec. Gempol, Cirebon, Jawa Barat, Indonesia      354   
17  Tj. Priok, Kota Jkt Utara, Daerah Khusus Ibuko...     15.6   
18  Tj. Priok, Kota Jkt Utara, Daerah Khusus Ibuko...       71   
19  Tj. Priok, Kota Jkt Utara, Daerah Khusus Ibuko...     52.6   

                                        originAddress  itemHeight itemValue  \
0           Bekasi, Bekasi City, West Java, Indonesia         1.0       0.0   
1   Tj. Priok, Kota Jkt Utara, Daerah Khusus Ibuko...         1.0       0.0   
2           Kec. Cikupa, Tangerang, Banten, Indonesia         1.0       0.0   
3           Kabupaten Karawang, Jawa Barat, Indonesia         1.0       0.0   
4   Kec. Batuceper, Kota Tangerang, Banten, Indonesia         1.0       0.0   
5   Kec. Batuceper, Kota Tangerang, Banten, Indonesia         1.0       0.0   
6   Kapuk, Kecamatan Cengkareng, Kota Jakarta Bara...         1.0       0.0   
7   Kecamatan Setiabudi, Kota Jakarta Selatan, Dae...         1.0       0.0   
8        Kec. Citeureup, Bogor, Jawa Barat, Indonesia         1.0       0.0   
9              Karawang Regency, West Java, Indonesia         1.0       0.0   
10                                                NaN         NaN       NaN   
11                                                NaN         NaN       NaN   
12   Jl. Pahlawan, Pekalongan, Jawa Tengah, Indonesia         1.0       0.0   
13  Kapuk, Kecamatan Cengkareng, Kota Jakarta Bara...         1.0       0.0   
14 

---
# 4. Import Dummy

This data contains list of cargo and truck company names that needs to be excluded while doing EDA process later.

In [ ]:
dummy = pd.read_excel('Data Dummy Perusahaan.xlsx')
dummy.head()

CompanyID                      Nama Perusahaan  \
0  1c4a8de8-d998-4e33-9032-59133b40bb35                        PT Test Cargo   
1  b192e134-3c2b-4e8f-ae86-8734229ac194              LOGEE TRANS CARGO OWNER   
2  4188588f-634b-4f5a-a318-af63f0d7092d                       PT. TEST TRUCK   
3  916ec6d3-af29-4f7d-ba90-0b0892d83e71  PT. PT TELEKOMUNIKASI INDONESIA TBK   
4  3e3dde23-e1c2-4929-a980-12ee3aa70056                  LOGEE TRANS TRUCKER   

  Cargo/Trucker? Unnamed: 3  
0          Cargo        NaN  
1          Cargo        NaN  
2        Trucker        NaN  
3        Trucker        NaN  
4        Trucker        NaN

---
# Data Manipulation

## 1. Filter Dummy

Disini kita akan menggunakan data dummy untuk mem-filter perusahaan pada data df_delivery.

In [ ]:
# Delete trailing space
dummy[['CompanyID','Nama Perusahaan']] = dummy[['CompanyID','Nama Perusahaan']].applymap(lambda x : x.strip())
# Upper companies name
dummy.loc[:,'Nama Perusahaan'] = dummy['Nama Perusahaan'].apply(lambda x : x.upper())

In [ ]:
dummy.head(10)

CompanyID                      Nama Perusahaan  \
0  1c4a8de8-d998-4e33-9032-59133b40bb35                        PT TEST CARGO   
1  b192e134-3c2b-4e8f-ae86-8734229ac194              LOGEE TRANS CARGO OWNER   
2  4188588f-634b-4f5a-a318-af63f0d7092d                       PT. TEST TRUCK   
3  916ec6d3-af29-4f7d-ba90-0b0892d83e71  PT. PT TELEKOMUNIKASI INDONESIA TBK   
4  3e3dde23-e1c2-4929-a980-12ee3aa70056                  LOGEE TRANS TRUCKER   
5  b19bb5b3-2215-440d-8c70-7a4be8e6bbdf                     PT. CLUB TRUCK Q   
6  17717fa7-839a-4289-87cc-9f5c3057238d               RENDI MAKMUR SEJAHTERA   
7  40579fcb-3218-4f79-b473-a35c3345eeed                            UNDEFINED   
8  7f507059-5a76-45b1-8c3d-a2d7f905f94a                              PT. AGI   
9  3bb917f7-fa36-4a1d-80cc-2608864c50c5                            PT. DUMMY   

  Cargo/Trucker? Unnamed: 3  
0          Cargo        NaN  
1          Cargo        NaN  
2        Trucker        NaN  
3        Trucker        NaN  
4        Trucker        NaN  
5        Trucker         QA  
6        Trucker        NaN  
7            NaN        NaN  
8        Trucker        NaN  
9        Trucker        NaN

## 2. Create List

Create a new list that contains companies name from dummy data.

In [ ]:
cargodummy = list(dummy['Nama Perusahaan'].unique())
new_cargodummy = []

for item in cargodummy:
    new_cargodummy.append(item)
display(new_cargodummy)
idummy = list(dummy['CompanyID'].unique())
display(idummy)

['PT TEST CARGO',
 'LOGEE TRANS CARGO OWNER',
 'PT. TEST TRUCK',
 'PT. PT TELEKOMUNIKASI INDONESIA TBK',
 'LOGEE TRANS TRUCKER',
 'PT. CLUB TRUCK Q',
 'RENDI MAKMUR SEJAHTERA',
 'UNDEFINED',
 'PT. AGI',
 'PT. DUMMY',
 'PT BERSAMA',
 'PT. TRANSGLOBAL INDONESIA ABADI CARGO',
 'PT. DUMY TRUCKER',
 '-',
 'PT. TESTING FLOU',
 'PT. TEST ICHA',
 'PT. DUMMY CARGO LTD',
 'PT. SURYA HAGANTA PRATAMA TRIAL',
 'PT SELALU JAYA',
 'CV. DUMMY TRUCK LTD',
 'PT. MULTITRADE PERKASA DUMMY',
 'PT. CLUB CARGO',
 'CV. BRAVO LOGISTICS',
 'PT. MEGATRADINDO',
 'PT. RAY CARGO',
 'PT. DUMMYYYYYYYYYYYYYYYYYY',
 'PT. CARGO PRODUCTION TESTING',
 'PT. TESTING CARGO',
 'PT. CARGO PORKOL',
 'RIAN RAHMAT SENTOSA',
 'PT. MANDIRI TRANSPORT DUMMY',
 'PT. DEV ABADI',
 'PT. RAX CARGO TRIAL',
 'PT. DUMYY',
 'PT. DUMMI DUMMI',
 'CV. FLOU TEST']

['1c4a8de8-d998-4e33-9032-59133b40bb35',
 'b192e134-3c2b-4e8f-ae86-8734229ac194',
 '4188588f-634b-4f5a-a318-af63f0d7092d',
 '916ec6d3-af29-4f7d-ba90-0b0892d83e71',
 '3e3dde23-e1c2-4929-a980-12ee3aa70056',
 'b19bb5b3-2215-440d-8c70-7a4be8e6bbdf',
 '17717fa7-839a-4289-87cc-9f5c3057238d',
 '40579fcb-3218-4f79-b473-a35c3345eeed',
 '7f507059-5a76-45b1-8c3d-a2d7f905f94a',
 '3bb917f7-fa36-4a1d-80cc-2608864c50c5',
 '96817a1a-a476-48cf-82e5-ce1c8ee8bd99',
 '0cbcff05-8dcd-4b95-9d3a-c69fc603a5e5',
 'bbcfe54e-306e-4ce5-8624-0f4038a83f0b',
 '17add240-988b-4bc6-bd1d-3e754dd29ceb',
 'ed04ac6a-910d-4eec-a2d0-de3565e2cd31',
 '4681cf8f-388b-47d7-9e7b-3219c8651d5d',
 '9edfe6a7-3ea0-49f9-a311-53828f7073d2',
 'ba1ffb66-c90e-4140-ae62-c1f86a3e93ca',
 '3d1c5b96-2f59-4e43-8430-ca3a989b7494',
 'fbb68258-16f7-41e8-a201-0a22673e8b74',
 'f365240f-303e-451d-b6c2-b1218502fa81',
 'cb4eb7b5-5ff3-4bab-ac93-ffaa84a48a54',
 'de98c7e5-18d6-4683-8c39-51dcb9f3a189',
 '555853de-12b5-45a9-8d5c-3f3730dc4b83',
 '1fdbf8eb-e47e-

## 3. Data Transformation 1

In this first data transformation, we will 

1. Replace null values in column cargoCompanyName and truckCompanyName into 'Other'.

2. Delete space at the start and end of every cargo and truck company name.

3. UPPER all company names.

In [ ]:
df_delivery['cargoCompanyName'].replace(np.nan,'Other',inplace=True)
df_delivery['truckCompanyName'].replace(np.nan,'Other',inplace=True)
df_delivery[['cargoCompanyName','truckCompanyName']] = df_delivery[['cargoCompanyName',
                                            'truckCompanyName']].applymap(lambda x : x.strip())
df_delivery[['cargoCompanyName','truckCompanyName']] = df_delivery[['cargoCompanyName',
                                            'truckCompanyName']].applymap(lambda x : x.upper())

In [ ]:
df_delivery[['cargoCompanyName','truckCompanyName']] 

cargoCompanyName       truckCompanyName
0                   OTHER                  OTHER
1           GPI LOGISTICS  GPI LOGISTICS TRUCKER
2      AVI CARGO SERVICES  AVI LOGISTICS TRUCKER
3      AVI CARGO SERVICES  AVI LOGISTICS TRUCKER
4      AVI CARGO SERVICES  AVI LOGISTICS TRUCKER
...                   ...                    ...
13732   KALDU SARI NABATI  SURYA PERSADA MANDIRI
13733   KALDU SARI NABATI  SURYA PERSADA MANDIRI
13734   KALDU SARI NABATI  SURYA PERSADA MANDIRI
13735   KALDU SARI NABATI  SURYA PERSADA MANDIRI
13736   KALDU SARI NABATI  SURYA PERSADA MANDIRI

[13737 rows x 2 columns]

## 4. Subset Data 1

Subset data of cargo and truck companies not in data dummy.

In [ ]:
df_delivery = df_delivery[~(df_delivery['cargoCompanyId'].isin(idummy))]
df_delivery = df_delivery[~(df_delivery['truckCompanyId'].isin(idummy))]

In [ ]:
df_delivery.shape

(13230, 99)

## 5. Transformasi Data 2

Because cargoCompanyName and truckCompanyName column have been filtered out from dummy, the next step is to revert 'Other' value back into its origin nan values.

In [ ]:
df_delivery['cargoCompanyName'].replace('Other', np.nan ,inplace=True)

In [ ]:
df_delivery['truckCompanyName'].replace('Other', np.nan ,inplace=True)

## 9. Delete Missing Values

In this task, we need to find the number of transactions made by each user of each cargo company. Therefore, the columns that are important to note are:

1. cargoCompanyId
2. cargoCompanyName
3. createdBy
4. totalAmount
5. tripFeeAmount
6. taxAmount

In this step, we will also check for null values in each of these columns.

In [ ]:
df_delivery[['cargoCompanyId','cargoCompanyName','createdBy','totalAmount','tripFeeAmount','taxAmount']].isna().sum()

cargoCompanyId      9
cargoCompanyName    0
createdBy           9
totalAmount         9
tripFeeAmount       9
taxAmount           9
dtype: int64

There are 9 empty rows of data in the cargoCompanyId, createdBy, totalAmount, tripFeeAmount, and taxAmount columns. Therefore, the next step is to retrieve records where there are no empty values ​​in all of the columns.



In [ ]:
delivs_notna = df_delivery.dropna(thresh=6,subset= ['cargoCompanyId','cargoCompanyName','createdBy',
                    'totalAmount','tripFeeAmount','taxAmount'])
delivs_notna.shape

(13221, 99)

## 10. Subset Data 2

Create new subset data that only consist of column cargoCompanyId, cargoCompanyName, createdBy, and totalAmount. 

In [ ]:
delivs_notna = delivs_notna[['cargoCompanyId','cargoCompanyName','createdBy','totalAmount']]
print(delivs_notna.shape)
display(delivs_notna)

(13221, 4)


cargoCompanyId    cargoCompanyName  \
1      bcefe6e0-0dd1-47db-8b7a-493a96f60aa4       GPI LOGISTICS   
2      6b326d3d-edb8-42ca-b07b-6db190a5d4b9  AVI CARGO SERVICES   
3      6b326d3d-edb8-42ca-b07b-6db190a5d4b9  AVI CARGO SERVICES   
4      6b326d3d-edb8-42ca-b07b-6db190a5d4b9  AVI CARGO SERVICES   
5      6b326d3d-edb8-42ca-b07b-6db190a5d4b9  AVI CARGO SERVICES   
...                                     ...                 ...   
13732  0d7f8fef-328b-4b75-a9ce-6e751a1bb82c   KALDU SARI NABATI   
13733  0d7f8fef-328b-4b75-a9ce-6e751a1bb82c   KALDU SARI NABATI   
13734  0d7f8fef-328b-4b75-a9ce-6e751a1bb82c   KALDU SARI NABATI   
13735  0d7f8fef-328b-4b75-a9ce-6e751a1bb82c   KALDU SARI NABATI   
13736  0d7f8fef-328b-4b75-a9ce-6e751a1bb82c   KALDU SARI NABATI   

                                  createdBy  totalAmount  
1      c25b016b-4b26-4fe5-9530-5f50aafee7ca    1815000.0  
2      29071794-8fd7-4d7d-ad83-5af348759bb4    1650000.0  
3      29071794-8fd7-4d7d-ad83-5af348759bb4    1980000.0  
4      29071794-8fd7-4d7d-ad83-5af348759bb4    1650000.0  
5      29071794-8fd7-4d7d-ad83-5af348759bb4    1650000.0  
...                                     ...          ...  
13732  09dce70a-a774-4d8e-aed7-096958800981    1800000.0  
13733  09dce70a-a774-4d8e-aed7-096958800981    7000000.0  
13734  09dce70a-a774-4d8e-aed7-096958800981    2500000.0  
13735  09dce70a-a774-4d8e-aed7-096958800981    2550000.0  
13736  09dce70a-a774-4d8e-aed7-096958800981    2550000.0  

[13221 rows x 4 columns]

In [ ]:
delivs_notna.reset_index(inplace=True)

In [ ]:
delivs_notna.drop('index',axis=1,inplace=True)

In [ ]:
delivs_notna

cargoCompanyId    cargoCompanyName  \
0      bcefe6e0-0dd1-47db-8b7a-493a96f60aa4       GPI LOGISTICS   
1      6b326d3d-edb8-42ca-b07b-6db190a5d4b9  AVI CARGO SERVICES   
2      6b326d3d-edb8-42ca-b07b-6db190a5d4b9  AVI CARGO SERVICES   
3      6b326d3d-edb8-42ca-b07b-6db190a5d4b9  AVI CARGO SERVICES   
4      6b326d3d-edb8-42ca-b07b-6db190a5d4b9  AVI CARGO SERVICES   
...                                     ...                 ...   
13216  0d7f8fef-328b-4b75-a9ce-6e751a1bb82c   KALDU SARI NABATI   
13217  0d7f8fef-328b-4b75-a9ce-6e751a1bb82c   KALDU SARI NABATI   
13218  0d7f8fef-328b-4b75-a9ce-6e751a1bb82c   KALDU SARI NABATI   
13219  0d7f8fef-328b-4b75-a9ce-6e751a1bb82c   KALDU SARI NABATI   
13220  0d7f8fef-328b-4b75-a9ce-6e751a1bb82c   KALDU SARI NABATI   

                                  createdBy  totalAmount  
0      c25b016b-4b26-4fe5-9530-5f50aafee7ca    1815000.0  
1      29071794-8fd7-4d7d-ad83-5af348759bb4    1650000.0  
2      29071794-8fd7-4d7d-ad83-5af348759bb4    1980000.0  
3      29071794-8fd7-4d7d-ad83-5af348759bb4    1650000.0  
4      29071794-8fd7-4d7d-ad83-5af348759bb4    1650000.0  
...                                     ...          ...  
13216  09dce70a-a774-4d8e-aed7-096958800981    1800000.0  
13217  09dce70a-a774-4d8e-aed7-096958800981    7000000.0  
13218  09dce70a-a774-4d8e-aed7-096958800981    2500000.0  
13219  09dce70a-a774-4d8e-aed7-096958800981    2550000.0  
13220  09dce70a-a774-4d8e-aed7-096958800981    2550000.0  

[13221 rows x 4 columns]

## 11. Import User Data

User data is required to get user name for every order.

In [ ]:
client = MongoClient("mongodb://localhost:27017")
db = client.get_database("logee_trans")
user = db.get_collection("user")

In [ ]:
df_user = pd.DataFrame(list(user.find({})))
print(df_user.shape)
display(df_user)

(2640, 21)


_id                                userId  \
0     606bdccd946a16117091a5b5  eb5dda16-612f-451d-b23d-4f316fe35cdd   
1     5f0ec5466e9c480010930374  f7fb9a6c-fcdf-45ac-8be1-3cfcc9d7743e   
2     5f28fc4bbea14b0010f94ce1  57a4b952-3aa7-4c51-8c76-5e38875a006c   
3     5f5f216f9710fa0012898130  db5c905b-b357-47fa-a4b4-55ceca1b0215   
4     5f5f23ec9710fa0012898132  b2c6042b-63c6-4e69-b8f0-1a45738d88ce   
...                        ...                                   ...   
2635  610a3b8b5144a80011a3ef3c  2996c678-fa36-42a5-b5db-6fd896e9a876   
2636  610a3c9b5144a80011a3ef3d  a2f55fc3-2d45-4184-b713-bd9b81724de5   
2637  610a3d5d5144a80011a3ef3e  fe06d6cc-3b98-4976-ac54-0465aee75469   
2638  610a3ed15144a80011a3ef3f  38d64230-0edd-4bf8-ae40-3c9a1a551364   
2639  610aaf515144a80011a3ef40  ac808dba-4b2e-452a-ba67-42c093b0613e   

                    modifiedAt  \
0     2021-08-05T03:00:50.542Z   
1     2020-07-16T10:23:15.983Z   
2     2020-08-04T06:16:09.903Z   
3     2020-09-14T07:53:19.864Z   
4     2021-04-19T08:34:42.449Z   
...                        ...   
2635  2021-08-04T07:02:35.878Z   
2636  2021-08-04T07:07:07.749Z   
2637  2021-08-04T07:10:21.367Z   
2638  2021-08-04T07:16:33.012Z   
2639  2021-08-04T15:18:09.051Z   

                                                 avatar  \
0                                                   NaN   
1     https://minio-asset.logeetrans.com/logee-trans...   
2     https://minio-asset.logeetrans.com/logee-trans...   
3     https://minio-asset.logeetrans.com/logee-trans...   
4     https://minio-asset.logeetrans.com/logee-trans...   
...                                                 ...   
2635  https://minio-asset.logeetrans.com/default/pla...   
2636  https://minio-asset.logeetrans.com/default/pla...   
2637  https://minio-asset.logeetrans.com/default/pla...   
2638  https://minio-asset.logeetrans.com/default/pla...   
2639  https://minio-asset.logeetrans.com/default/pla...   

                       name                            email  \
0                       NaN                              NaN   
1            Truck Reviewer      truck.reviewer2@yopmail.com   
2     Ilman Naafian Wirawan         ilman.naafianw@gmail.com   
3                    rahman        hanifaelfiana20@gmail.com   
4              Jono Purhadi         g.adi.prasetyo@gmail.com   
...                     ...                              ...   
2635                SYAFRIL                                    
2636                GUSRIAL                                    
2637                M JAINI                                    
2638                 JASNO                                     
2639          Dani Hermawan  dani.hermawan@kalogistics.co.id   

                                        apps                       roles  \
0                                        NaN                         NaN   
1     [transport-truck, transport-dashboard]  [truck-owner, truck-admin]   
2     [transport-cargo, transport-dashboard]  [cargo-owner, cargo-admin]   
3     [transport-truck, transport-dashboard]  [truck-owner, truck-admin]   
4     [transport-cargo, transport-dashboard]  [cargo-owner, cargo-admin]   
...                                      ...                         ...   
2635                      [transport-driver]                    [driver]   
2636                      [transport-driver]                    [driver]   
2637                      [transport-driver]                    [driver]   
2638                      [transport-driver]                    [driver]   
2639  [transport-cargo, transport-dashboard]  [cargo-owner, cargo-admin]   

     isActive isDeleted                 createdAt  \
0         NaN       NaN                       NaN   
1        True     False  2020-07-15T08:58:46.504Z   
2        True     False  2020-08-04T06:12:27.606Z   
3       False     False  2020-09-14T07:53:19.864Z   
4        True      True  2020-09-14T08:03:56.076Z   
...       ...       ...                     

## 12. Create List

Create a new list that contains pairs of user name and its corresponding user id.

In [ ]:
users = df_user[['userId','name']]
pengguna = []
for x,y in zip(list(users.userId.values), list(users.name.values)):
    pengguna.append((x,y))
pengguna

[('eb5dda16-612f-451d-b23d-4f316fe35cdd', nan),
 ('f7fb9a6c-fcdf-45ac-8be1-3cfcc9d7743e', 'Truck Reviewer'),
 ('57a4b952-3aa7-4c51-8c76-5e38875a006c', 'Ilman Naafian Wirawan'),
 ('db5c905b-b357-47fa-a4b4-55ceca1b0215', 'rahman'),
 ('b2c6042b-63c6-4e69-b8f0-1a45738d88ce', 'Jono Purhadi'),
 ('30df7b60-d39c-4437-be49-fcf0dcb41115', 'TATANG UMBARA'),
 ('8a386abf-49a5-4b6f-b1b3-b0d549ddc54d', 'PT SPARK'),
 ('72367cf4-d229-4d5d-b891-944281028fe0', 'IDRIS'),
 ('f3fda1ac-68a2-431e-a456-faedfe4189ce', 'IWAN HERMAWAN'),
 ('5e1f7801-beb2-4b8e-aa29-3eb09fa278db', 'ADE ANSORI'),
 ('878d14b0-f5b4-4598-a2b1-b7399c6ee576', 'PT ALIMAS CIPTA SEJAHTERA'),
 ('becdfe3a-405e-4180-9c38-4813100f1fd5', 'Ummi Robbaniyah Mushthofa'),
 ('0974fbb7-bc8c-444c-a74e-4893b366bf9f', 'Fandy Aditya'),
 ('cdda9822-cf44-49fb-9fde-55353ed02de4', 'Danny langkun'),
 ('18e0d5e8-dc08-44a1-a362-575076a9c546', 'Citarasa Prima '),
 ('0a8b612c-fcc5-401f-a1b6-944900fcf52c', 'ADAY S'),
 ('3f137399-03d3-4fc0-bb1a-90c76ec0bba5', 'BASRI 

In [ ]:
len(pengguna)

2640

## 13. Data Transformation 3

In [ ]:
# Function to get name from every id in createdBy column
def name(x):
    for ids,name in pengguna:
        if x == ids:
            return name
        else :
            pass
delivs_notna.loc[:,'name'] = delivs_notna.createdBy.apply(lambda x : name(x))

In [ ]:
delivs_notna.head(1000)

cargoCompanyId             cargoCompanyName  \
0    bcefe6e0-0dd1-47db-8b7a-493a96f60aa4                GPI LOGISTICS   
1    6b326d3d-edb8-42ca-b07b-6db190a5d4b9           AVI CARGO SERVICES   
2    6b326d3d-edb8-42ca-b07b-6db190a5d4b9           AVI CARGO SERVICES   
3    6b326d3d-edb8-42ca-b07b-6db190a5d4b9           AVI CARGO SERVICES   
4    6b326d3d-edb8-42ca-b07b-6db190a5d4b9           AVI CARGO SERVICES   
..                                    ...                          ...   
995  bcefe6e0-0dd1-47db-8b7a-493a96f60aa4                GPI LOGISTICS   
996  5e8fa145-feb7-44ba-ba5a-11c39fccffa2  FESA ANTARAN LOGISTIK CARGO   
997  5e8fa145-feb7-44ba-ba5a-11c39fccffa2  FESA ANTARAN LOGISTIK CARGO   
998  6b326d3d-edb8-42ca-b07b-6db190a5d4b9           AVI CARGO SERVICES   
999  6b326d3d-edb8-42ca-b07b-6db190a5d4b9           AVI CARGO SERVICES   

                                createdBy  totalAmount  \
0    c25b016b-4b26-4fe5-9530-5f50aafee7ca    1815000.0   
1    29071794-8fd7-4d7d-ad83-5af348759bb4    1650000.0   
2    29071794-8fd7-4d7d-ad83-5af348759bb4    1980000.0   
3    29071794-8fd7-4d7d-ad83-5af348759bb4    1650000.0   
4    29071794-8fd7-4d7d-ad83-5af348759bb4    1650000.0   
..                                    ...          ...   
995  becdfe3a-405e-4180-9c38-4813100f1fd5    1800000.0   
996  ee009167-4acf-43b4-97da-4a2eb2013365    6000000.0   
997  ee009167-4acf-43b4-97da-4a2eb2013365    6500000.0   
998  29071794-8fd7-4d7d-ad83-5af348759bb4    1800000.0   
999  29071794-8fd7-4d7d-ad83-5af348759bb4    1500000.0   

                          name  
0                         dede  
1                          ina  
2                          ina  
3                          ina  
4                          ina  
..                         ...  
995  Ummi Robbaniyah Mushthofa  
996                      Admin  
997                      Admin  
998                        ina  
999                        ina  

[1000 rows x 5 columns]

Values in totalAmount column are still in float. Convert it into integer.

In [ ]:
delivs_notna['totalAmount'] = delivs_notna['totalAmount'].astype(np.int64)

In [ ]:
delivs_notna

cargoCompanyId    cargoCompanyName  \
0      bcefe6e0-0dd1-47db-8b7a-493a96f60aa4       GPI LOGISTICS   
1      6b326d3d-edb8-42ca-b07b-6db190a5d4b9  AVI CARGO SERVICES   
2      6b326d3d-edb8-42ca-b07b-6db190a5d4b9  AVI CARGO SERVICES   
3      6b326d3d-edb8-42ca-b07b-6db190a5d4b9  AVI CARGO SERVICES   
4      6b326d3d-edb8-42ca-b07b-6db190a5d4b9  AVI CARGO SERVICES   
...                                     ...                 ...   
13216  0d7f8fef-328b-4b75-a9ce-6e751a1bb82c   KALDU SARI NABATI   
13217  0d7f8fef-328b-4b75-a9ce-6e751a1bb82c   KALDU SARI NABATI   
13218  0d7f8fef-328b-4b75-a9ce-6e751a1bb82c   KALDU SARI NABATI   
13219  0d7f8fef-328b-4b75-a9ce-6e751a1bb82c   KALDU SARI NABATI   
13220  0d7f8fef-328b-4b75-a9ce-6e751a1bb82c   KALDU SARI NABATI   

                                  createdBy  totalAmount               name  
0      c25b016b-4b26-4fe5-9530-5f50aafee7ca      1815000               dede  
1      29071794-8fd7-4d7d-ad83-5af348759bb4      1650000                ina  
2      29071794-8fd7-4d7d-ad83-5af348759bb4      1980000                ina  
3      29071794-8fd7-4d7d-ad83-5af348759bb4      1650000                ina  
4      29071794-8fd7-4d7d-ad83-5af348759bb4      1650000                ina  
...                                     ...          ...                ...  
13216  09dce70a-a774-4d8e-aed7-096958800981      1800000  Kaldusari Nabati   
13217  09dce70a-a774-4d8e-aed7-096958800981      7000000  Kaldusari Nabati   
13218  09dce70a-a774-4d8e-aed7-096958800981      2500000  Kaldusari Nabati   
13219  09dce70a-a774-4d8e-aed7-096958800981      2550000  Kaldusari Nabati   
13220  09dce70a-a774-4d8e-aed7-096958800981      2550000  Kaldusari Nabati   

[13221 rows x 5 columns]

In [ ]:
delivs_notna.sort_values(by='cargoCompanyName',ascending=True,inplace=True)

In [ ]:
delivs_notna

cargoCompanyId         cargoCompanyName  \
7767   847309dd-5a12-4116-8aa6-63b77e8686aa        AGUNG SERBA TULEN   
8840   847309dd-5a12-4116-8aa6-63b77e8686aa        AGUNG SERBA TULEN   
9494   847309dd-5a12-4116-8aa6-63b77e8686aa        AGUNG SERBA TULEN   
10940  847309dd-5a12-4116-8aa6-63b77e8686aa        AGUNG SERBA TULEN   
8075   847309dd-5a12-4116-8aa6-63b77e8686aa        AGUNG SERBA TULEN   
...                                     ...                      ...   
8545   b2fdefd0-ff71-4739-8f4a-1b5320cdbb2b    VARUNA TIRTA PRAKASYA   
3619   b2fdefd0-ff71-4739-8f4a-1b5320cdbb2b    VARUNA TIRTA PRAKASYA   
9099   b2fdefd0-ff71-4739-8f4a-1b5320cdbb2b    VARUNA TIRTA PRAKASYA   
6599   d42b0974-0837-4e64-ac29-33780bf2b622  ZAIN EKPRESS BERSAUDARA   
2657   63dc286f-ff56-42bc-8943-f571ebdc46bd   ZEESINDO INTERNASIONAL   

                                  createdBy  totalAmount  \
7767   9a587c17-d6e7-41fa-b411-e700d91d254c     20000000   
8840   9a587c17-d6e7-41fa-b411-e700d91d254c     28000000   
9494   9a587c17-d6e7-41fa-b411-e700d91d254c     32800000   
10940  9a587c17-d6e7-41fa-b411-e700d91d254c     22500000   
8075   9a587c17-d6e7-41fa-b411-e700d91d254c     28000000   
...                                     ...          ...   
8545   a56f9f05-4127-4b34-b5d0-2a306b880bf3      2000000   
3619   3751dbe9-a5db-4bc5-814a-be733c49e4d6      1000000   
9099   a56f9f05-4127-4b34-b5d0-2a306b880bf3      2000000   
6599   82876aee-2c2b-42ad-acab-985fef4666f7      1300000   
2657   a79df818-0f7e-4746-98a6-88f538e59529      5700000   

                            name  
7767        PT Agung Serba Tulen  
8840        PT Agung Serba Tulen  
9494        PT Agung Serba Tulen  
10940       PT Agung Serba Tulen  
8075        PT Agung Serba Tulen  
...                          ...  
8545      Habibie Albi Althoyibi  
3619                Andy Pratama  
9099      Habibie Albi Althoyibi  
6599             Vedro Fernandes  
2657   PT ZEESINDO INTERNASIONAL  

[13221 rows x 5 columns]

Delete unnecessary space in cargoCompanName column.

In [ ]:
delivs_notna.loc[:,'cargoCompanyName'] = delivs_notna.cargoCompanyName.apply(lambda x : x.strip())
delivs_notna.sort_values(by='cargoCompanyName',ascending=True,inplace=True)
delivs_notna

cargoCompanyId         cargoCompanyName  \
7767   847309dd-5a12-4116-8aa6-63b77e8686aa        AGUNG SERBA TULEN   
10350  847309dd-5a12-4116-8aa6-63b77e8686aa        AGUNG SERBA TULEN   
10656  847309dd-5a12-4116-8aa6-63b77e8686aa        AGUNG SERBA TULEN   
10367  847309dd-5a12-4116-8aa6-63b77e8686aa        AGUNG SERBA TULEN   
7213   847309dd-5a12-4116-8aa6-63b77e8686aa        AGUNG SERBA TULEN   
...                                     ...                      ...   
3572   b2fdefd0-ff71-4739-8f4a-1b5320cdbb2b    VARUNA TIRTA PRAKASYA   
6932   b2fdefd0-ff71-4739-8f4a-1b5320cdbb2b    VARUNA TIRTA PRAKASYA   
6966   b2fdefd0-ff71-4739-8f4a-1b5320cdbb2b    VARUNA TIRTA PRAKASYA   
6599   d42b0974-0837-4e64-ac29-33780bf2b622  ZAIN EKPRESS BERSAUDARA   
2657   63dc286f-ff56-42bc-8943-f571ebdc46bd   ZEESINDO INTERNASIONAL   

                                  createdBy  totalAmount  \
7767   9a587c17-d6e7-41fa-b411-e700d91d254c     20000000   
10350  9a587c17-d6e7-41fa-b411-e700d91d254c     27500000   
10656  9a587c17-d6e7-41fa-b411-e700d91d254c     35000000   
10367  9a587c17-d6e7-41fa-b411-e700d91d254c     17500000   
7213   9a587c17-d6e7-41fa-b411-e700d91d254c     35000000   
...                                     ...          ...   
3572   3751dbe9-a5db-4bc5-814a-be733c49e4d6      1800000   
6932   a56f9f05-4127-4b34-b5d0-2a306b880bf3      2480000   
6966   a56f9f05-4127-4b34-b5d0-2a306b880bf3      2480000   
6599   82876aee-2c2b-42ad-acab-985fef4666f7      1300000   
2657   a79df818-0f7e-4746-98a6-88f538e59529      5700000   

                            name  
7767        PT Agung Serba Tulen  
10350       PT Agung Serba Tulen  
10656       PT Agung Serba Tulen  
10367       PT Agung Serba Tulen  
7213        PT Agung Serba Tulen  
...                          ...  
3572                Andy Pratama  
6932      Habibie Albi Althoyibi  
6966      Habibie Albi Althoyibi  
6599             Vedro Fernandes  
2657   PT ZEESINDO INTERNASIONAL  

[13221 rows x 5 columns]

## 14. Data Grouping

In this step, the data will be grouped based on cargoCompanyName, name, and createdBy columns. After that, we do sum and count on totalAmount column. This is a mandatory step because data grouping requires aggregation process on data.

In [ ]:
delivery_final = delivs_notna.groupby(['cargoCompanyName','name','createdBy']).agg({'totalAmount': ['sum','count']})
delivery_final

totalAmount  \
                                                                                                        sum   
cargoCompanyName                 name                      createdBy                                          
AGUNG SERBA TULEN                PT Agung Serba Tulen      9a587c17-d6e7-41fa-b411-e700d91d254c  3072300000   
ALAM INDAH ANUGERAH              dellciamarotama           f80f05ed-e6fd-4273-8371-6754a27c26a2   399359909   
AMANAH AKHLAK MULIA              PT Amanah Akhlak Mulia    e793efa4-d312-4e3a-a63f-dc6d0a995f7f   936000000   
ANDILO NAJOGI                    adend                     783aaa55-6316-4ca7-a55c-a4d7202ed240  1530700000   
ANTASENA JASA                    antasena                  e62f3215-c7e3-45bf-ae5e-1830ffa6ff19   235453100   
...                                                                                                     ...   
TUJUH BERKAH SEMESTA CARGO OWNER chandra                   cc6add4f-fa8a-4396-ac9b-0dc96352cf4e    11500000   
VARUNA TIRTA PRAKASYA            Andy Pratama              3751dbe9-a5db-4bc5-814a-be733c49e4d6     8800000   
                                 Habibie Albi Althoyibi    a56f9f05-4127-4b34-b5d0-2a306b880bf3    47920000   
ZAIN EKPRESS BERSAUDARA          Vedro Fernandes           82876aee-2c2b-42ad-acab-985fef4666f7     1300000   
ZEESINDO INTERNASIONAL           PT ZEESINDO INTERNASIONAL a79df818-0f7e-4746-98a6-88f538e59529     5700000   

                                                                                                       
                                                                                                count  
cargoCompanyName                 name                      createdBy                                   
AGUNG SERBA TULEN                PT Agung Serba Tulen      9a587c17-d6e7-41fa-b411-e700d91d254c   117  
ALAM INDAH ANUGERAH              dellciamarotama           f80f05ed-e6fd-4273-8371-6754a27c26a2   374  
AMANAH AKHLAK MULIA              PT Amanah Akhlak Mulia    e793efa4-d312-4e3a-a63f-dc6d0a995f7f    36  
ANDILO NAJOGI                    adend                     783aaa55-6316-4ca7-a55c-a4d7202ed240   130  
ANTASENA JASA                    antasena                  e62f3215-c7e3-45bf-ae5e-1830ffa6ff19    62  
...                                                                                               ...  
TUJUH BERKAH SEMESTA CARGO OWNER chandra                   cc6add4f-fa8a-4396-ac9b-0dc96352cf4e     3  
VARUNA TIRTA PRAKASYA            Andy Pratama              3751dbe9-a5db-4bc5-814a-be733c49e4d6     7  
                                 Habibie Albi Althoyibi    a56f9f05-4127-4b34-b5d0-2a306b880bf3    23  
ZAIN EKPRESS BERSAUDARA          Vedro Fernandes           82876aee-2c2b-42ad-acab-985fef4666f7     1  
ZEESINDO INTERNASIONAL           PT ZEESINDO INTERNASIONAL a79df818-0f7e-4746-98a6-88f538e59529     1  

[111 rows x 2 columns]

In [ ]:
delivery_final.reset_index(inplace=True)
delivery_final

cargoCompanyName                       name  \
                                                                   
0                   AGUNG SERBA TULEN       PT Agung Serba Tulen   
1                 ALAM INDAH ANUGERAH            dellciamarotama   
2                 AMANAH AKHLAK MULIA     PT Amanah Akhlak Mulia   
3                       ANDILO NAJOGI                      adend   
4                       ANTASENA JASA                   antasena   
..                                ...                        ...   
106  TUJUH BERKAH SEMESTA CARGO OWNER                    chandra   
107             VARUNA TIRTA PRAKASYA               Andy Pratama   
108             VARUNA TIRTA PRAKASYA     Habibie Albi Althoyibi   
109           ZAIN EKPRESS BERSAUDARA            Vedro Fernandes   
110            ZEESINDO INTERNASIONAL  PT ZEESINDO INTERNASIONAL   

                                createdBy totalAmount        
                                                  sum count  
0    9a587c17-d6e7-41fa-b411-e700d91d254c  3072300000   117  
1    f80f05ed-e6fd-4273-8371-6754a27c26a2   399359909   374  
2    e793efa4-d312-4e3a-a63f-dc6d0a995f7f   936000000    36  
3    783aaa55-6316-4ca7-a55c-a4d7202ed240  1530700000   130  
4    e62f3215-c7e3-45bf-ae5e-1830ffa6ff19   235453100    62  
..                                    ...         ...   ...  
106  cc6add4f-fa8a-4396-ac9b-0dc96352cf4e    11500000     3  
107  3751dbe9-a5db-4bc5-814a-be733c49e4d6     8800000     7  
108  a56f9f05-4127-4b34-b5d0-2a306b880bf3    47920000    23  
109  82876aee-2c2b-42ad-acab-985fef4666f7     1300000     1  
110  a79df818-0f7e-4746-98a6-88f538e59529     5700000     1  

[111 rows x 5 columns]

In [ ]:
delivery_final.columns = ['cargoCompanyName','name','Id','Total_transaction','Frequency']

In [ ]:
delivery_final

cargoCompanyName                       name  \
0                   AGUNG SERBA TULEN       PT Agung Serba Tulen   
1                 ALAM INDAH ANUGERAH            dellciamarotama   
2                 AMANAH AKHLAK MULIA     PT Amanah Akhlak Mulia   
3                       ANDILO NAJOGI                      adend   
4                       ANTASENA JASA                   antasena   
..                                ...                        ...   
106  TUJUH BERKAH SEMESTA CARGO OWNER                    chandra   
107             VARUNA TIRTA PRAKASYA               Andy Pratama   
108             VARUNA TIRTA PRAKASYA     Habibie Albi Althoyibi   
109           ZAIN EKPRESS BERSAUDARA            Vedro Fernandes   
110            ZEESINDO INTERNASIONAL  PT ZEESINDO INTERNASIONAL   

                                       Id  Total_transaction  Frequency  
0    9a587c17-d6e7-41fa-b411-e700d91d254c         3072300000        117  
1    f80f05ed-e6fd-4273-8371-6754a27c26a2          399359909        374  
2    e793efa4-d312-4e3a-a63f-dc6d0a995f7f          936000000         36  
3    783aaa55-6316-4ca7-a55c-a4d7202ed240         1530700000        130  
4    e62f3215-c7e3-45bf-ae5e-1830ffa6ff19          235453100         62  
..                                    ...                ...        ...  
106  cc6add4f-fa8a-4396-ac9b-0dc96352cf4e           11500000          3  
107  3751dbe9-a5db-4bc5-814a-be733c49e4d6            8800000          7  
108  a56f9f05-4127-4b34-b5d0-2a306b880bf3           47920000         23  
109  82876aee-2c2b-42ad-acab-985fef4666f7            1300000          1  
110  a79df818-0f7e-4746-98a6-88f538e59529            5700000          1  

[111 rows x 5 columns]

Now, grouped data again based on cargoCompanyName, name, Id, Total_transaction, and Frequency columns.

In [ ]:
delivery_final_group = delivery_final.groupby(['cargoCompanyName','name','Id','Total_transaction','Frequency']).count()
delivery_final_group

Empty DataFrame
Columns: []
Index: [(AGUNG SERBA TULEN, PT Agung Serba Tulen, 9a587c17-d6e7-41fa-b411-e700d91d254c, 3072300000, 117), (ALAM INDAH ANUGERAH, dellciamarotama, f80f05ed-e6fd-4273-8371-6754a27c26a2, 399359909, 374), (AMANAH AKHLAK MULIA, PT Amanah Akhlak Mulia, e793efa4-d312-4e3a-a63f-dc6d0a995f7f, 936000000, 36), (ANDILO NAJOGI, adend, 783aaa55-6316-4ca7-a55c-a4d7202ed240, 1530700000, 130), (ANTASENA JASA, antasena, e62f3215-c7e3-45bf-ae5e-1830ffa6ff19, 235453100, 62), (ANTASENA JASA, fiqih fajrin, 7e379ad5-f61d-4273-b770-6aa1e47656af, 747283250, 232), (ANTASENA JASA, ita, ff7965fe-96ae-44e3-a034-e0d6bb9666c4, 134153200, 37), (ANTASENA JAYA, ita, ff7965fe-96ae-44e3-a034-e0d6bb9666c4, 1650000, 1), (AVI CARGO SERVICES, Admin, ee009167-4acf-43b4-97da-4a2eb2013365, 41100000, 30), (AVI CARGO SERVICES, Admin Logeetrans Cargo, db924c77-ce29-4be4-9611-f5dd3bd32dd7, 9000000, 6), (AVI CARGO SERVICES, Benni Nofrizal, 2213794b-de7f-44b8-8545-4f5569d30e36, 14225000, 3), (AVI CARGO SERVICES, FITRIA UTAMI, 0ab96482-89e3-4998-9441-24b4b1b3ce96, 104000000, 9), (AVI CARGO SERVICES, Mellysa Icha, 3749dafc-7f88-4656-890f-ed174321ee84, 62460000, 42), (AVI CARGO SERVICES, PT Wildan kaya abadi, df905981-83ca-4790-8928-b52a185e9caf, 71000000, 46), (AVI CARGO SERVICES, Raffi Ahmad, 012a6300-5776-439d-b41a-973979bdb545, 60500000, 52), (AVI CARGO SERVICES, devina, acbbdc2e-00e9-4674-aa04-920fe83e354a, 88150000, 67), (AVI CARGO SERVICES, ina, 29071794-8fd7-4d7d-ad83-5af348759bb4, 581960000, 398), (BCG CARGO, Muhammad Fadhiila, 91df0c57-0aa3-44cd-94e1-b0f4407094b6, 21500000, 11), (BERLIAN TRANSTAR ABADI, BTSA Zepri, 6a33931e-7320-4d2b-a413-f54ab1f380d7, 11200000, 7), (BHANDA CAHYA UTAMA, PT Bandha Cahya Utama, a85a1c49-e528-4998-b40c-dca32f288c8f, 4000000, 2), (BHANDA GHARA REKSA CARGO, PT Bhanda Ghara Reksa , c92c4131-6650-4510-8c47-8e084d8c4eb4, 146000000, 5), (BIFORST LOGISTICS, Reni Fajarwati, 63d94895-5cae-4d83-bf7e-6cb6bc91e56b, 24150000, 8), (BUANA FARAS SAKTI, PT Buans faras Sakti , fc762d6b-cbb2-42d6-acfc-ecb9aa06715f, 67900000, 24), (BUMI MANFAAT GEMILANG, PT Bumi Manfaat Gemilang, bd4c48e7-66ed-4f28-a3ca-736f1d8a7ccb, 8723000000, 1282), (CARMA GUNA MANDIRI CARGO, PT Carma Guna Mandiri, d1a2800d-ed4f-456e-9f6a-20c5beaf82a3, 1384500000, 87), (CAROLINE TRANS MANDIRI, PTCaroline Trans Mandiri, 1ed52d02-974a-4779-b869-b6ba436c8c01, 63000000, 2), (CIPTA KRIDA BAHARI, kintetsu, ecf8a3c4-a5fc-4fab-baa0-19dc6f57f99b, 63400000, 15), (CITARASA PRIMA INDONESIA BERJAYA, Citarasa Prima , 18e0d5e8-dc08-44a1-a362-575076a9c546, 7316450000, 1343), (DAGANG MESIN MANDIRI INDONESIA, Yudhie Lesmana, aa646dea-4558-48c2-b56e-0c33843f26e0, 6234900000, 295), (DAMRI LOGISTICS CARGO, Operational, 4528bc2d-5639-420f-a57c-064601ca69a7, 9650000, 2), (DELLCIA MORATAMA NUSANTARA CARGO, Dellcia Moratama, ea081341-5268-4964-b439-850faed5034b, 1800000, 1), (DELLCIA MORATAMA NUSANTARA CARGO, dellciamarotama, f80f05ed-e6fd-4273-8371-6754a27c26a2, 357000000, 12), (DUA PULUH LOGISTIK CARGO, supriyadi, 66e2e266-8948-4c89-9a9a-17f9bd462951, 1200000, 1), (FESA ANTARAN LOGISTIK CARGO, Admin, ee009167-4acf-43b4-97da-4a2eb2013365, 1880000000, 310), (FESA ANTARAN LOGISTIK CARGO, Admin Logeetrans Cargo, db924c77-ce29-4be4-9611-f5dd3bd32dd7, 42800000, 7), (FESA ANTARAN LOGISTIK CARGO, Mellysa Icha, 3749dafc-7f88-4656-890f-ed174321ee84, 672454000, 123), (FESA ANTARAN LOGISTIK CARGO, Raffi Ahmad, 012a6300-5776-439d-b41a-973979bdb545, 249600000, 47), (FESA ANTARAN LOGISTIK CARGO, Suryono fesa, fae2b07a-24fc-47c8-8ee9-ddc2b833c6ce, 17000000, 3), (FESA ANTARAN LOGISTIK CARGO, echi fesa, 27cda829-8cc3-4e4d-8da6-62806b511591, 535000000, 119), (FESA ANTARAN LOGISTIK CARGO, ina, 29071794-8fd7-4d7d-ad83-5af348759bb4, 21100000, 4), (GEMILANG SUKSES MANDIRI, Angga Torikho, 5cc95858-afea-42a6-9d99-30ca4311464f, 526000000, 24), (GLAGAHWANGI BARWASO INDO, Gelagah Wangi, 0c4eb664-0494-4ea1-bd5b-cbd46dfa6d00, 5945020000, 1025), (GPI LOGISTICS, Admin, ee009167-4acf-43b4-97da-4a2eb2013365, 36300000

# Insights

---
## 1. Total of Minimum and Maximum Transaction

### a. Top 10 Users with Highest Total Transaction

In [ ]:
delivery_final_group.sort_values(by = 'Total_transaction',ascending = False).head()

Empty DataFrame
Columns: []
Index: [(KALDU SARI NABATI, Kaldusari Nabati , 09dce70a-a774-4d8e-aed7-096958800981, 8767168000, 1663), (BUMI MANFAAT GEMILANG, PT Bumi Manfaat Gemilang, bd4c48e7-66ed-4f28-a3ca-736f1d8a7ccb, 8723000000, 1282), (CITARASA PRIMA INDONESIA BERJAYA, Citarasa Prima , 18e0d5e8-dc08-44a1-a362-575076a9c546, 7316450000, 1343), (JAYA BERSAMA, CV Jaya Bersama , 1bd104d4-eb22-457e-b3d2-ee1d34202bbc, 6468680000, 555), (DAGANG MESIN MANDIRI INDONESIA, Yudhie Lesmana, aa646dea-4558-48c2-b56e-0c33843f26e0, 6234900000, 295)]

### b. Top 10 Users with Lowest Total Transaction

In [ ]:
delivery_final_group.sort_values(by = 'Total_transaction',ascending = True).head()

Empty DataFrame
Columns: []
Index: [(METRO PASUNDAN PRATAMA CARGO, Rizki alhakim, 9653b5b8-b0b5-4320-a71d-d45eff447cac, 484000, 1), (JASA PRIMA LOGISTIK BULOG, ahmad zuraijun, 8c550414-256c-4abf-9c85-fb5d7bcbf8ea, 500000, 1), (INDAH BUANA REZEKI CARGO, indah puspanegara, 4f2581fe-9581-4ceb-91ad-ee2ded675a7d, 1000000, 1), (DUA PULUH LOGISTIK CARGO, supriyadi, 66e2e266-8948-4c89-9a9a-17f9bd462951, 1200000, 1), (ZAIN EKPRESS BERSAUDARA, Vedro Fernandes, 82876aee-2c2b-42ad-acab-985fef4666f7, 1300000, 1)]

Based on the final data above, it was found that user named Rizki Alhakim from the cargo company METRO PASUNDAN PRATAMA CARGO had the smallest total number of transactions, which was around Rp. 484,000 and only made 1 transaction. Meanwhile, a user named Kaldusari Nabati from the cargo company KALDU SARI NABATI had the largest total transaction, which was Rp. 8,767,168,000 and has made as many as 1663 transactions.

In [ ]:
# Export to excel
#delivfixgrup.to_excel('cargoTransactionew.xlsx')

## 2. Visualization

In [ ]:
delivery = delivery_final_group.reset_index()
delivery

cargoCompanyName                       name  \
0                   AGUNG SERBA TULEN       PT Agung Serba Tulen   
1                 ALAM INDAH ANUGERAH            dellciamarotama   
2                 AMANAH AKHLAK MULIA     PT Amanah Akhlak Mulia   
3                       ANDILO NAJOGI                      adend   
4                       ANTASENA JASA                   antasena   
..                                ...                        ...   
106  TUJUH BERKAH SEMESTA CARGO OWNER                    chandra   
107             VARUNA TIRTA PRAKASYA               Andy Pratama   
108             VARUNA TIRTA PRAKASYA     Habibie Albi Althoyibi   
109           ZAIN EKPRESS BERSAUDARA            Vedro Fernandes   
110            ZEESINDO INTERNASIONAL  PT ZEESINDO INTERNASIONAL   

                                       Id  Total_transaction  Frequency  
0    9a587c17-d6e7-41fa-b411-e700d91d254c         3072300000        117  
1    f80f05ed-e6fd-4273-8371-6754a27c26a2          399359909        374  
2    e793efa4-d312-4e3a-a63f-dc6d0a995f7f          936000000         36  
3    783aaa55-6316-4ca7-a55c-a4d7202ed240         1530700000        130  
4    e62f3215-c7e3-45bf-ae5e-1830ffa6ff19          235453100         62  
..                                    ...                ...        ...  
106  cc6add4f-fa8a-4396-ac9b-0dc96352cf4e           11500000          3  
107  3751dbe9-a5db-4bc5-814a-be733c49e4d6            8800000          7  
108  a56f9f05-4127-4b34-b5d0-2a306b880bf3           47920000         23  
109  82876aee-2c2b-42ad-acab-985fef4666f7            1300000          1  
110  a79df818-0f7e-4746-98a6-88f538e59529            5700000          1  

[111 rows x 5 columns]

### Visualization Total Amount of Transaction Per User

In [ ]:
delivery.iplot(kind='bar',x='name',y='Total_transaction',color='blue',theme='solar');

In [ ]:
delivery.iplot(kind='bar',x='name',y='Frequency',color='red',theme='solar');